### PR TITLE
Automate Epinio-server update for the chart epinio

### PR DESCRIPTION
**Pull request by olblak**
# Automate Epinio-server update for the chart epinio

## Description
This "pullrequest" introduces two changes. 
.1 ~~Parametrize container image and tag~~
.2 Automatically open a pullrequest when a new epinio server is available

### ~~Parametrize container image and tag~~

~~By parameterizing the container image and the tag used in the chart epinio, we can easily use the same chart but with a different container in testing environment like with the following command~~

### Enable updatecli workflow
Updatecli is a tool to define a custom update pipeline.
The purpose is to automatically open a pullrequest the chart need to be updated with the latest epinio-server version
In this pipeline, I propose the following workflow"

* Retrieve the latest release from epinio/epinio
* Test that a docker image "splatform/epinio:<latest-release>" has been published on dockerhub
* Test that "splatform/epinio:<latest-release>" is the container defined in the chart epinio
* When container tag mismatch between the chart and the epinio-server version, then it commits the new version to a temporary branch.
* Open a pullrequest with the labels "dependencies" and "epinio" from a temporary branch to the main one.

! Please note that the current updatecli configuration depends on this [PR](https://github.com/updatecli/updatecli/pull/388) to be merged first, which should be done in the coming days.
Therefore it won't work with the latest updatecli stable release

---
Here is an example of the updatecli executed locally

<details><summary>Click to expand</summary>

```
updatecli diff --config updatecli/updatecli.d/epinio.yaml --values updatecli/values.yaml                                                                                                 ✔ 


+++++++++++
+ PREPARE +
+++++++++++

Loading Pipeline "updatecli/updatecli.d/epinio.yaml"
Repository retrieved: 1


++++++++++++
+ PIPELINE +
++++++++++++



##############################
# BUMP EPINIO SERVER VERSION #
##############################


SOURCES
=======

epinio
------
Searching for version matching pattern "latest"
✔ Github Release version "v0.2.1" found matching pattern "latest"


CHANGELOG:
----------

Release published on the 2021-11-22 14:47:54 +0000 UTC at the url https://github.com/epinio/epinio/releases/tag/v0.2.1

**Implemented enhancements:**

- Try to integrate the Epinio Helm Chart in the Rancher Catalog [\#976](https://github.com/epinio/epinio/issues/976)
- Persist information on application deployment method [\#912](https://github.com/epinio/epinio/issues/912)
- Empty list from API should return `\[\]` instead of `null` [\#436](https://github.com/epinio/epinio/issues/436)
- Create/manage extended namespace names for epinio organizations [\#434](https://github.com/epinio/epinio/issues/434)
- `epinio install` now needs a system-domain . It won't try to use a "magic" domain anymore (used to be `omg.howdoi.website`).
Users can still use a "magic" domain if they want to but it has to be explicitly passed to the install command  [\#983](https://github.com/epinio/epinio/issues/983)

**Fixed bugs:**

- Make API responses compatible with the JSON API Spec [\#806](https://github.com/epinio/epinio/issues/806)
- Subsequent image builds for the same app are stored in different image names [\#991](https://github.com/epinio/epinio/issues/991)
- Uninstall breaks when using external S3 storage [\#990](https://github.com/epinio/epinio/issues/990)
- Improved API documentation [\#321](https://github.com/epinio/epinio/issues/321)


CONDITIONS:
===========

dockerImageContainer
--------------------
✔ Key "container.image", in YAML file "chart/epinio/values.yaml", is correctly set to "splatform/epinio-server"

dockerImage
-----------
✔ splatform/epinio-server:v0.2.1 available on the Docker Registry


TARGETS
========

epinio-helm-charts
------------------

**Dry Run enabled**

INFO: Using spec.Value instead of source input value.
✔ Key 'container.tag', from file 'chart/epinio/values.yaml', already set to v0.2.1, nothing else need to be done


PULL REQUESTS
=============

=============================

REPORTS:


✔ EPINIO.YAML:
	Sources:
		✔ [epinio] Get Latest epinio version(githubRelease)
	Condition:
		✔ [dockerImage] Check that splatform/epinio-installer:{{ source "epinio" }} is published(dockerImage)
		✔ [dockerImageContainer] Check that splatform/epinio-installer is used in Helm chart epinio(yaml)
	Target:
		✔ [epinio-helm-charts]  Update epinio version in chart epinio(helmChart)



Run Summary
===========
Pipeline(s) run:
  * Changed:	0
  * Failed:	0
  * Skipped:	0
  * Succeeded:	1
  * Total:	1
```

</details>

### CODE REVIEW

**Comment by olblak** [comment]
@juadk I added a second pipeline to bump the chart epinio-installer when a new version of the chart epinio is available. I still have a bunch of improvements, hence the commented part.

File: chart/epinio/templates/server.yaml [hunk 1]
```diff
@@ -238,7 +238,7 @@ spec:
             - name: REGISTRY_CERTIFICATE_SECRET
               value: "{{ .Values.registry.certificateSecret }}"
             {{- end }}
-          image: splatform/epinio-server:{{ .Chart.AppVersion }}
+          image: {{ .Values.container.image }}:{{ .Values.container.tag }}
```
Comment (manno) [inline chart/epinio/templates/server.yaml:241]: This might as well go into a separate PR. Epinio needs to support the Rancher style for specifying images.
Example: https://github.com/rancher/charts/blob/dev-v2.6/charts/longhorn/longhorn/1.1.100/templates/deployment-driver.yaml#L23

https://github.com/epinio/epinio/issues/979

**Review by manno — CHANGES_REQUESTED** [review]

**Comment by olblak** [comment]
:/ I messed up the git history :p

File: chart/epinio/templates/server.yaml (hunk 1 above)
Comment (olblak) [reply chart/epinio/templates/server.yaml:241]: After evaluation, I am wondering if that extra complexity is needed at this stage of the chart. Because it doesn't make sense to reuse the same structure for the values file defined on your link, and excepted the image published on dockerhub by docker like bash, we only  use the splatform/epinio-server  which could be overridden by just updatecli the value of .Values.container.image

Comment (olblak) [reply chart/epinio/templates/server.yaml:241]: Dimitri covered this in hise pullrequest https://github.com/epinio/helm-charts/pull/38

**Comment by olblak** [comment]
The current pullrequest is on hold until this is finalized https://github.com/epinio/helm-charts/pull/38

**Comment by olblak** [comment]
This pullrequest is ready to be merged (@manno, @jimmykarily ).
The updatecli Github Action will be triggered in one of the following scenarios

1. Manually triggered
2. After an Epinio release
3. Once a week

It will open a pull request when a change needs to be applied.

Specifically related to an Epinio release, it will open a pullrequest containing updates for the Epinio chart, bumping the chart  Version with a minor increment, the app version set to the latest Epinio version, and bumping the container tag in values.yaml with the latest version.

File: updatecli/values.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+github:
+  epinio:
+    user: "GitHub Actions"
+    email: "41898282+github-actions[bot]@users.noreply.github.com"
```
Comment (jimmykarily) [inline updatecli/values.yaml:4]: what does this value match to?

Comment (olblak) [reply updatecli/values.yaml:4]: That's the "user" who commits using the github_token and github_actor
An example can be found [here](https://github.com/jenkins-infra/jenkins-infra/pull/2007)

File: .github/workflows/updatecli.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+jobs:
+  updatecli:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Apply
+        uses: updatecli/updatecli-action@v1.17.0
+        if: github.ref == 'refs/heads/main'
```
Comment (manno) [inline .github/workflows/updatecli.yml:25]: I'd move that to `jobs.updatecli.if`, so the checkout step is not run

File: .github/workflows/updatecli.yml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+---
+name: updatecli
```
Comment (manno) [inline .github/workflows/updatecli.yml:2]: I like that the name matches the filename. However, I wonder if we can find something that describes what the workflow actually does ("create pr in xy repo", "bump x in y repo"). On the other hand, a comment might be enough.

File: updatecli/updatecli.d/epinio.yaml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    name: "Check that splatform/epinio-installer is used in Helm chart epinio"
+    kind: "yaml"
+    disablesourceinput: true
+    scmID: "helm-charts"
+    spec:
+      file: "chart/epinio/values.yaml"
+      key: "image.epinio.repository"
+      value: "splatform/epinio-server"
+
+# Defines what need to be udpated if needed
```
Comment (manno) [inline updatecli/updatecli.d/epinio.yaml:58]:
```suggestion
# Defines what needs to be udpated if needed
```
Not sure about this comment. Does this create the PR, together with the values from `pullrequest` above?

**Review by manno — APPROVED** [review]

File: .github/workflows/updatecli.yml (hunk 3 above)
Comment (olblak) [reply .github/workflows/updatecli.yml]: Thanks for the very good suggestion
